### PR TITLE
Backend: Implement history service

### PR DESCRIPTION
**Pull request by ekweirui**
Currently, our project does not have the functionality for N2: History: Maintain a record of the questions attempted by the user e.g., maintain a list of questions attempted along with the date-time of attempt, the attempt itself and/or suggested solutions.

This needs to be changed so as to meet our project requirements of having five nice to haves.

Thus, this pull request aims to do the following:
* Implement the backend history service to achieve N2
* Standardize the frontend calls to the backend history service

### CODE REVIEW

File: frontend/src/services/QuestionHistoryFunctions.ts
Comment (Tan-Jia-Rong) [inline frontend/src/services/QuestionHistoryFunctions.ts:1]: brutal eh AHAHAHA jk

File: backend/history_service/controllers/historyController.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        const usersRef = db.collection("users").doc(uid);
+        const userDoc = await usersRef.get();
+        
+        if (!userDoc.exists) {
+            return res.status(404).json({ message: "User not found" });
+        }
+
+        // Retrieve and filter questionsAttempted array
+        let questionsAttempted = userDoc.data().questionsAttempted || [];
+        const cutoffDate = moment().tz("Asia/Singapore").subtract(180, 'days'); // 180 days ago in Singapore time
```
Comment (kenvynKwek) [inline backend/history_service/controllers/historyController.js:17]: amazing work 🔥 might have to test the api again after the other branches merge in

**Review by kenvynKwek — APPROVED** [review]
LVGTM 🔥 thanks for the amazing hard work ! bless ur soul

File: backend/history_service/middleware/authenticateToken.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      if (response.ok) {
+        const userData = await response.json(); 
+        req.user = userData; // Store user data in request
+        next();
+      } else {
+        const errorData = await response.json(); // Get the error response
+        res.status(response.status).json({ message: errorData.message || 'Invalid Token' });
+      }
+    } catch (error) {
+      res.status(500).json({ message: 'Internal Question Service Server Error' });
```
Comment (Tan-Jia-Rong) [inline backend/history_service/middleware/authenticateToken.js:28]: Question Service?

File: backend/history_service/controllers/historyController.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        // Create a new array with the updated entry
+        const updatedQuestionsAttempted = [
+            ...questionsAttempted.slice(0, index), // Entries before the index
+            { 
+                ...questionsAttempted[index], // Keep the existing fields
+                codeWritten: codeWritten // Update the codeWritten
+            },
+            ...questionsAttempted.slice(index + 1) // Entries after the index
+        ];
```
Comment (Tan-Jia-Rong) [inline backend/history_service/controllers/historyController.js:142]: low key inefficient but i guess it will do since we decided on an array

**Review by Tan-Jia-Rong — COMMENTED** [review]
generally LGTM just some small comments

**Pull request merged**